### ISSUE TITLE
Make gives linker error 

### ISSUE THREAD

**Issue by dineyw23** [rethinkdb/rethinkdb#5754]
Fetched latest commits from source and getting the following error while building on Mac OSX El Capitan. 
```
Undefined symbols for architecture x86_64:
  "_CRYPTO_THREADID_set_callback", referenced from:
      crypto::initialization_guard_t::initialization_guard_t() in initialization_guard.o
      crypto::initialization_guard_t::initialization_guard_t() in initialization_guard.o
  "_CRYPTO_THREADID_set_numeric", referenced from:
      crypto::initialization_guard_t::initialization_guard_t()::$_0::__invoke(crypto_threadid_st*) in initialization_guard.o
  "_PKCS5_PBKDF2_HMAC", referenced from:
      crypto::detail::pbkcs5_pbkdf2_hmac_sha256(char const*, unsigned long, unsigned char const*, unsigned long, unsigned int) in pbkcs5_pbkdf2_hmac.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [build/release_clang/rethinkdb] Error 1
make: *** [make] Error 2
```
Is it related to my Open SSL linking or version?
```command
openssl version -a

OpenSSL 1.0.2h  3 May 2016
built on: reproducible build, date unspecified
platform: darwin64-x86_64-cc
options:  bn(64,64) rc4(ptr,int) des(idx,cisc,16,int) idea(int) blowfish(idx) 
compiler: clang -I. -I.. -I../include  -fPIC -fno-common -DOPENSSL_PIC -DZLIB_SHARED -DZLIB -DOPENSSL_THREADS -D_REENTRANT -DDSO_DLFCN -DHAVE_DLFCN_H -arch x86_64 -O3 -DL_ENDIAN -Wall -DOPENSSL_IA32_SSE2 -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_MONT5 -DOPENSSL_BN_ASM_GF2m -DSHA1_ASM -DSHA256_ASM -DSHA512_ASM -DMD5_ASM -DAES_ASM -DVPAES_ASM -DBSAES_ASM -DWHIRLPOOL_ASM -DGHASH_ASM -DECP_NISTZ256_ASM
OPENSSLDIR: "/usr/local/etc/openssl"
```

**Comment by danielmewes** [comment]
Thanks for reporting this @dineyw23 .

This is weird, since 1) OpenSSL 1.0.2 should be compatible as far as I know and 2) It's failing during linking rather than compiling, which means that the missing functions were present in the OpenSSL headers but not in the actual library.

Maybe it's linking against a different OpenSSL library than what it's compiling with. I can see that OpenSSL 1.0.2h is installed in `/usr/local` on your system.

@dineyw23 The easiest way to work around this is probably adding `--fetch openssl` to your `./configure` flags and then recompiling.

@larkost Could it be that we're using the OpenSSL includes from /usr/local/include because the OpenSSL that comes with OS X doesn't have any include files, and then link against the system OpenSSL because it comes earlier in the library path?

**Comment by dineyw23** [comment]
@danielmewes Thanks. Adding `--fetch openssl` flag to `./configure` worked perfect.

**Comment by larkost** [comment]
@danielmewes: I think you have the problem diagnosed correctly: the OpenSSL 1.0.2 headers are the only one in the resolution path (because El Capitan removes the system OpenSSL headers), but the binary for the system version is ahead of the `/usr/local` in the resolution order, so it grabs that one.

Since this sort of case is more possible since Apple removed those headers, I wonder if we should try anything to make sure that things match? There are going to be lots of edge cases, but it might prevent some problems.

**Comment by danielmewes** [comment]
Could we add a special case for OS X in the configure script / Makefile to make sure that we ignore the system's libcrypto? Is there even a syntax for telling the linker to use a specific copy of a library?

**Comment by larkost** [comment]
Reading the `ld` manpage (on linux) I found that if we specify the folders where the versions of libraries that we want with the `-L` flag, those should be searched first, before the default ones. I will make sure that this is also the case on MacOS, and then put that into our link flags to manipulate the search order.

**Comment by larkost** [comment]
After thinking about this a bit more I think that I was on the wrong track to solve this specific issue, but that they are related. I think that this is what is happening in this case:

1. In this case there are two versions of OpenSSL installed on the system: Apple's at `/usr/lib/libssl.0.9.dylib` (0.9.8zh) and a user-installed version probably in `/usr/local/lib` (1.0.2h).
2. As of El Capitan Apple's version no longer installs headers, so the only headers on the system belong to the user-installed version.
3. When we go to compile the compiler is checking its normal paths, and so finds the only headers available (1.0.2h)
4. But the system dylib (0.9.8.zh) is on the path before the user-installed version, so it will use that one.

So the solution is to add `-L /usr/local/lib/openssl` (or something like that) to `LDFLAGS`. That should cause the linker to do the right thing. Unfortunately, I don't think that there is any way that we can automate that because without real magic (read: heavy, and fragile scripting) there is no way of associating specific headers with a specific library.

We can however improve our error detection on this by enhancing the [testing dummy program](https://github.com/rethinkdb/rethinkdb/blob/dbbef604536324d789e4903ecb0f1be68e361f56/configure#L920-L925) in `./configure` to make sure we can both compile and link against the specific things we need from later versions of OpenSSL.

Over to @VeXocide who has agreed to do that.

**Comment by VeXocide** [comment]
This is in CR 3742 by @danielmewes.

**Comment by VeXocide** [comment]
Merged into `next` via commit d3347d60cb06565472a22f788f333fe6d28848a8.
